### PR TITLE
Add JsonSerializerOptions.MakeReadOnly(bool) overload.

### LINKED ISSUE #89934: Provide an API for configuring `JsonSerializerOptions` instances for reflection-based serialization.
## Background & Motivation

The System.Text.Json reflection-based serializer makes it possible for users to supply `JsonSerializerOptions` instances that don't have a `TypeInfoResolver` property explicitly configured. It does this by silently populating the `TypeInfoResolver` property with a reflection-based `DefaultJsonTypeInfoResolver` value and then locking the instance for further modification.

Using publicly available APIs, this semantic can be emulated as follows:
```C#
void ConfigureForReflection(JsonSerializerOptions options)
{
    options.TypeInfoResolver ??= new DefaultJsonTypeInfoResolver();
    options.MakeReadOnly();
}
```
The problem with this approach is that it isn't thread safe when applied to the same `options` instance. I encountered this issue when attempting to implement reflection-based initialization in `System.Net.Http.Json` (cf. https://github.com/dotnet/runtime/issues/89830), which can only access public STJ APIs. At the moment, the only viable workaround for libraries like `System.Net.Http.Json` is to do the following:
```C#
if (options.TypeInfoResolver is null)
{
    // Run a basic serialization operation to force thread-safe initialization of the options instance.
    JsonSerializer.Deserialize<int>("0"u8, options);
}
```
While the above works, it is kind of wasteful in that it forces an unnecessary serialization operation. We need access to the API that simply primes the options instance for reflection-based usage.

## API Proposal

Add a method to `JsonSerializerOptions` for configuring using reflection-based defaults:

```diff
namespace System.Text.Json;

public partial class JsonSerializerOptions
{
    // Throws IOE if TypeInfoResolver is null
    public void MakeReadOnly();

+   // Populates TypeInfoResolver with DefaultJsonTypeInfoResolver if null
+   // (or the empty resolver if the IsReflectionEnabledDefault switch is disabled)
+   [RequiresUnreferencedCode, RequiresDynamicCode]
+   public void MakeReadOnlyWithReflectionDefaults();
}
```

cc @eerhardt

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Fix #89934.

cc @halter73 @eerhardt

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs [hunk 1]
```diff
@@ -45,11 +45,14 @@ public JsonConverter GetConverter(Type typeToConvert)
                 ThrowHelper.ThrowArgumentNullException(nameof(typeToConvert));
             }
 
-            if (JsonSerializer.IsReflectionEnabledByDefault && _typeInfoResolver is null)
+            if (JsonSerializer.IsReflectionEnabledByDefault)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:48]: Unrelated to the other changes, but it appears that the previous method structure resulted in rooted reflection when the feature switch is disabled in the trimming tests. Is this a known issue of the linker?

cc @vitek-karas @eerhardt

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/TrimmingTests/IsReflectionEnabledByDefaultFalse.cs [hunk 2]
```diff
@@ -44,18 +44,41 @@ public static int Main()
         {
         }
 
+        // Calling GetConverter should throw NotSupportedException.
+        try
+        {
+            _ = options.GetConverter(typeof(MyPoco));
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/System.Text.Json.Tests/TrimmingTests/IsReflectionEnabledByDefaultFalse.cs:50]: @eerhardt @vitek-karas for context to https://github.com/dotnet/runtime/pull/90013/files#r1284314974 adding this method resulted in rooting the reflection resolver.

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs [hunk 3]
```diff
@@ -402,6 +402,9 @@ public sealed partial class JsonSerializerOptions
         public System.Text.Json.Serialization.JsonConverter GetConverter(System.Type typeToConvert) { throw null; }
         public System.Text.Json.Serialization.Metadata.JsonTypeInfo GetTypeInfo(System.Type type) { throw null; }
         public void MakeReadOnly() { }
+        [System.Diagnostics.CodeAnalysis.RequiresDynamicCodeAttribute("Populating unconfigured TypeInfoResolver properties with the reflection resolver requires runtime code generation.")]
+        [System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute("Populating unconfigured TypeInfoResolver properties with the reflection resolver requires unreferenced code.")]
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/ref/System.Text.Json.cs:406]: Should we augment the message to say if you're passing false you should use the other overload instead?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs [hunk 4]
```diff
@@ -45,11 +45,14 @@ public JsonConverter GetConverter(Type typeToConvert)
...
             {
                 // Backward compatibility -- root & query the default reflection converters
                 // but do not populate the TypeInfoResolver setting.
-                return DefaultJsonTypeInfoResolver.GetConverterForType(typeToConvert, this);
+                if (_typeInfoResolver is null)
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:52]: was the lack of this check a bug previously?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs [hunk 5]
```diff
@@ -680,18 +678,21 @@ internal bool CanUseFastPathSerializationLogic
...
         /// A <see cref="JsonSerializerOptions"/> instance can be locked either if
         /// it has been passed to one of the <see cref="JsonSerializer"/> methods,
         /// has been associated with a <see cref="JsonSerializerContext"/> instance,
-        /// or a user explicitly called the <see cref="MakeReadOnly"/> method on the instance.
+        /// or a user explicitly called the <see cref="MakeReadOnly()"/> methods on the instance.
+        ///
+        /// Read-only instances use caching when querying <see cref="JsonConverter"/> and <see cref="JsonTypeInfo"/> metadata.
         /// </remarks>
         public bool IsReadOnly => _isReadOnly;
+        private volatile bool _isReadOnly;
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:692]: Why did you move this?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs (hunk 4 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:52]: See https://github.com/dotnet/runtime/pull/90013#discussion_r1284314974. Seems the previous structure was resulting in rooted reflection components in trimmed apps even though it's equivalent.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs (hunk 5 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:692]: Seems reasonable to collocate. It was previously grouped together with the fields that were being touched by the copy constructor but this one isn't.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs [hunk 6]
```diff
@@ -751,11 +786,18 @@ internal void ConfigureForJsonSerializer()
                 ThrowHelper.ThrowInvalidOperationException_JsonSerializerIsReflectionDisabled();
             }
 
-            MakeReadOnly();
+            Debug.Assert(_typeInfoResolver != null);
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:789]: This method is reading _typeInfoResolver multiple times. What happens if another thread sets it in the interim? I'm on my phone so haven't reviewed all the code, but how is this all thread-safe?

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs (hunk 3 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/ref/System.Text.Json.cs:406]: I thought about it, but it felt like cramming two separate concerns in the same warning. Most uses would be passing `true` and the warnings are still relevant there.

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/TrimmingTests/IsReflectionEnabledByDefaultFalse.cs (hunk 2 above)
Comment (vitek-karas) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests/TrimmingTests/IsReflectionEnabledByDefaultFalse.cs:50]: Would you have a branch/test we can look at?
I agree that the feature switch should make this work, so the fact that it doesn't is interesting.
(Adding @sbomer as I'm going to be OOF next week)

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs (hunk 6 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:789]: It generally speaking isn't, for example if you try to run `JsonSerializer.Serialize(x, options);` and `options.TypeInfoResolver = null;` calls concurrently that particular assertion will fail. We've been regarding that particular condition as acceptable, if you try to both configure an options instance and serialize with it concurrently you should expect undefined behavior. 

What should be thread-safe though is concurrent calls to `JsonSerializer.Serialize(x, options);` which is I wasn't able to achieve in the `System.Net.Http.Json` layer.

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/TrimmingTests/IsReflectionEnabledByDefaultFalse.cs (hunk 2 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests/TrimmingTests/IsReflectionEnabledByDefaultFalse.cs:50]: Sure, here's one: https://github.com/dotnet/runtime/commit/b2ba402683d34c58cd486814904b6391ed954703

**Review by stephentoub — APPROVED** [review]

**Review by tarekgh — APPROVED** [review]

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs [hunk 7]
```diff
@@ -706,11 +707,45 @@ public void MakeReadOnly()
...
+        /// </remarks>
+        [RequiresUnreferencedCode("Populating unconfigured TypeInfoResolver properties with the reflection resolver requires unreferenced code.")]
+        [RequiresDynamicCode("Populating unconfigured TypeInfoResolver properties with the reflection resolver requires runtime code generation.")]
+        public void MakeReadOnly(bool populateMissingResolver)
+        {
+            if (populateMissingResolver)
+            {
+                if (!_isConfiguredForJsonSerializer)
+                {
+                    ConfigureForJsonSerializer();
```
Comment (eerhardt) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:732]: ConfigureForJsonSerializer() will throw an exception when `JsonSerializer.IsReflectionEnabledByDefault == false && TypeInfoResolver == null` saying:

```
Reflection-based serialization has been disabled for this application. Either use the source generator APIs or explicitly configure the 'JsonSerializerOptions.TypeInfoResolver' property.
```

I don't think ASP.NET Core will be able to use this new API because of this. There are scenarios where we need to call `MakeReadOnly()` on a JsonSerializerOptions that doesn't have its TypeInfoResolver set yet. One example is in MVC:

https://github.com/dotnet/aspnetcore/issues/49360

MVC always creates a SystemTextJsonOutputFormatter up front without knowing if the app is actually going to use JSON serialization or not. When the app doesn't touch JSON, we shouldn't be throwing an exception at startup.

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:732]: It looks like aspnetcore was able to work around https://github.com/dotnet/runtime/issues/89830 since it owns its `JsonSerializerOptions` instances, so we don't need new APIs to unblock that case. 

The semantics of the new overload is oriented towards components accepting `JsonSerializerOptions` instances they don't own but still need to emulate the behavior of the `JsonSerializer` methods.

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/TrimmingTests/IsReflectionEnabledByDefaultFalse.cs (hunk 2 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests/TrimmingTests/IsReflectionEnabledByDefaultFalse.cs:50]: I just chanced on this discussion again, has this been looked at? I could create an issue if necessary.
